### PR TITLE
[RISCV] Introduce VLOptimizer pass

### PR DESCRIPTION
**Pull request by michaelmaitland**
The purpose of this optimization is to make the VL argument, for instructions that have a VL argument, as small as possible. This is implemented by visiting each instruction in reverse order and checking that if it has a VL argument, whether the VL can be reduced.

By putting this pass before VSETVLI insertion, we see three kinds of changes to generated code:
1. Eliminate VSETVLI instructions
2. Reduce the VL toggle on VSETVLI instructions that also change vtype
3. Reduce the VL set by a VSETVLI instruction

The list of supported instructions is currently whitelisted for safety. In the future, we could add more instructions to `isSupportedInstr` to support even more VL optimization.

We originally wrote this pass because vector GEP instructions do not take a VL, which leads us to emit code that uses VL=VLMAX to implement GEP in the RISC-V backend. As a result, some of the vector instructions will write to lanes, specifically between the intended VL and VLMAX, that will never be read. As an alternative to this pass, we considered adding a vector predicated GEP instruction, but this would not fit well into the intrinsic type system since GEP has a variable number of arguments, each with arbitrary types. The second approach we considered was to put this pass after VSETVLI insertion, but we found that it was more difficult to recognize optimization opportunities, especially across basic block boundaries -- the data flow analysis was also a bit more expensive and complex.

While this pass solves the GEP problem, we have expanded it to handle more cases of VL optimization, and there is opportunity for the analysis to be improved to enable even more optimization. We have a few follow up patches to post, but figured this would be a good start.

Co-authored-by: Craig Topper <craig.topper@sifive.com>
Co-authored-by: Kito Cheng <kito.cheng@sifive.com>

### CODE REVIEW

**Comment by preames** [comment]
Hey, awesome!  Really happy to see this making it's way upstream.   As I'm sure you know, I've been slowly expanding tryToReduceVL in RISCVVectorPeephole in this direction.  We should definitely make sure we're sharing as much code as possible.  I think we still need the early peephole (to drive the vmv.v.v conversions), but we should at least share code.

I'm going to take a real look at the code on Monday, but wanted to state initial support!

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (wangpc-pp) [inline llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:269]: Maybe we can reuse some parts of https://github.com/llvm/llvm-project/pull/105945?

**Review by wangpc-pp — COMMENTED** [review]
Great! I just left some comments. I will have a detailed review later.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1410]: In RISCVVectorPeephole we're modelling this with

```cpp
  bool ElementsDependOnVL = RISCVII::elementsDependOnVL(
      TII->get(RISCV::getRVVMCOpcode(Src->getOpcode())).TSFlags);
  if (ElementsDependOnVL || Src->mayRaiseFPException())
    return false;
```

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:269]: This is great that the EEW is modelled for every operand now. Either we could extract getOperandInfo out into somewhere reusable like RISCVInstrInfo and remove DestEEW. Or we could somehow also move this into TableGen. I wonder if we could do something similar to schedule classes where we can annotate the EMUL/EEW for each operand? Something like

```
  def V  : VALUVV<funct6, OPMVV, opcodestr # "." # vw # "v">,
           SchedBinaryMC<"WriteVIWALUV", "ReadVIWALUV", "ReadVIWALUV">,
           EEW<SEWx2, SEW, SEW>, EMUL<LMULx2, LMUL, LMUL>;
```

Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:269]: Sounds great! I think the TableGen way can be feasible and useful!

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1410]: Would you like to see this change in this PR? My concern is that supporting more instructions is going to require testing. It might be easier to take this change for a minimal number of instructions to make sure the pass is how everyone wants it, and then to have follow up patches which add this support + tests for that support.

File: llvm/test/CodeGen/RISCV/rvv/vl-opt-no-prop.ll
Comment (preames) [inline llvm/test/CodeGen/RISCV/rvv/vl-opt-no-prop.ll:1]: These three test files seem to have valuable coverage regardless of which approach we take towards reducing VL.  Can you commit these separately and rebase?

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1448]: Visiting each user for each time we visit something on the worklist (combined with an extra visit from the outer loop), seems potentially expensive.  We could restructure this along the lines of a instcombine style worklist which integrates the outer pass.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1455]: I think there might be a bug here.  The reduction still needs the VL to be non-zero.  If all other users are a register which happens to be zero at runtime, I think this can result in value reaching the reduction (which does not itself have the VL reduced) with a wrong scalar input value.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1369]: A better name for the way this is used might be: canReadPastVL.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1488]: The choice to only handle registers here is a surprising limitation.  We can likely generalize this for immediates as well.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1508]: This can be hoisted out of the loop right?  Though actually, how does such a thing end up on the visit list at all?  This maybe implies a missing guard somewhere.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (preames) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1430]: Please don't update for now.  We can generalize in separate patches after this goes in.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:206]: This silent off by one on the indexing is confusing to say the least.  Given this is only used in a handful of places, I'd prefer to see this inlined and specialized for each user.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:687]: I believe all of these hard coded LMUL on the whole register stuff are equal to MIVLMul.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:711]: You should be able to drive these off the recently added TableGen flag for DestEEW.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (preames) [inline llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1193]: I think this (and the other reduction) results is wrong.  Specifically, the destination has scalar LMUL, but the source has something else.  The fact we're returning the same regardless of op number here can't be correct.

**Review by preames — COMMENTED** [review]
A couple of detail comments inline, but I want to focus on high level review strategy.

Looking at the code for this, I spot a couple possible bugs (flagged in inline comments), but nothing really terrible code structure wise.  I'd like to propose that we defer most of the code cleanup suggestions until after this is in tree.  Specifically, I'm proposing the following:
1) This patch is repurposed to add the transform pass off-by-default.  We subset as needed to get something correct in tree with minimal delay.
2) We iterate in tree to common code, and to add back any key features subset in step (1).  
3) We enable the pass in it's own commit.  

I'm very hesitant to do the usual style review here because there's so much room for subtle bugs to be introduced, and if we get into such a cycle the review could extend nearly indefinitely.  I'm willing to assume good faith follow up here, and frankly will likely invest a bunch of time myself in getting this code into long term supportable state.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1488]: We have this generalization written, was going to post as a follow up.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1193]: From the RVV spec:

> The unsigned vwredsumu.vs instruction zero-extends the SEW-wide vector elements before summing them, then adds the 2*SEW-width scalar element, and stores the result in a 2*SEW-width scalar element.

> The vwredsum.vs instruction sign-extends the SEW-wide vector elements before summing them.

> The scalar input and output operands are held in element 0 of a single vector register, not a vector register group, so any vector register can be the scalar source or destination of a vector reduction regardless of LMUL setting.

Although the result is a single scalar, it's still written to a vector register. We can see here that the RVV spec refers to the elements it writes to have a SEW. I think it makes less sense to talk about LMUL here, and I think that I said to use the same relation that the SEW had for simplicity. As used, I don't think it would cause any harm since the LMUL being reported is larger than what is actually getting used (just a single element of a vector register).

Truthfully, I cant recall why I wrote it such that 14.1 instructions  return Unknown, but 14.2-14.4 don't. Some ideas:
* return Unknown
* keep as is
* modify OperandInfo to describe this case

I probably prefer the first option for now and we can always improve it in the future. Thoughts?

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:687]: We don't have pseudo instructions for whole register load/store and we don't appear to set the TSFlags `VLMul` for non-pseudo instructions.

**Comment by michaelmaitland** [comment]
As I am working on responding to reviews, one thing has become clear. We need better tests for this. I'm not sure what the best way to test things are. It feels like there are two things here:

1. getOperandInfo
2. the actual VL optimization

I wonder if it would be a good idea to add unit tests for `getOperandInfo`. We would go through each operand of each instruction and verify that it had the expected OperandInfo.

Then for the actual VL optimization, I think it would be best to use lit tests to get code coverage of all paths not (not including all paths through getOperandInfo since that would be tested on its own).

Does anyone have different ideas or does this sound okay?

**Comment by preames** [comment]
> Does anyone have different ideas or does this sound okay?

First, I want to explicitly say again that I'd welcome this iteration happening in tree.

Second, I don't tend to be a huge fan of C++ unit tests for this kind of API because they end up just repeating the implementation, and are hard to update.  If you want to go in this direction, maybe something along the lines of a printer pass so that diffs can be autogenerated and we only have to (manually) change one place?

Third, part of the problem with getOperandInfo is that it replicates information in other places.  In particular, I think this gets a lot simpler if restructured using the recently added DestEEW mechanism.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (preames) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:687]: Ok, another idea...

Their LMUL should match the register class right?  We don't have fractional register classes, but we also don't have fraction whole register moves?  So maybe combine this with isRVVWholeLoadStore from RISCVInstrInfo.cpp for the selection, then use the register class?

(I'm just looking for ways to remove redundancy.  If we can share code with other users, that increases the odds we'd catch a bug.)

Again, this is minor, and I'd be happy to iterate on this type of topic in tree.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1511]: At first glance, I agreed with the statement that we can hoist this out. But under closer inspection, I am not so sure.

> We've already checked above that UserOp is a vector register.

I think this part is key. If we hoist this check and continue in the outer loop, then we are not sure that UserOp is a vector register. I have to give some thought to whether we'd continue if both were non vector reg class.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1458]: I don't know if I follow your thinking here.

On a side note, ff the VL is zero, we cannot reduce it any further. We shouldn't even be calling tryReduceVL.

**Comment by michaelmaitland** [comment]
> First, I want to explicitly say again that I'd welcome this iteration happening in tree.

If everyone is on board for this approach, I think that would be good. But I want to highlight here that I think the test coverage is not in a place where I'd feel comfortable to enabling this pass by default. I think this is a must have before turning it on.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1432]: Sorry I should have clarified, I am happy to white-list instructions in this PR. I mainly wanted to point out the link between the two passes, and that we should cross-check RISCVVectorPeephole to make sure we've covered the cases mentioned in the comment above!

**Comment by lukel97** [comment]
> If everyone is on board for this approach, I think that would be good. But I want to highlight here that I think the test coverage is not in a place where I'd feel comfortable to enabling this pass by default. I think this is a must have before turning it on.

I would be happy to iterate in tree as well, we can leave it off by default and add some .mir tests. As for moving getOperandInfo out into TableGen, it was mostly a refactoring idea and I don't think we should block this PR on it. I can take a stab at it this week in parallel and see if it's feasible or not.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1458]: > If all other users are a register which happens to be zero at runtime, I think this can result in value reaching the reduction (which does not itself have the VL reduced) with a wrong scalar input value.

All the users have to have the same VL, so the reduction's VL would also be zero and it wouldn't read the scalar input value I think.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp]: I wrote in 1c35bd52fbd466528e4ab073045531762469de7f to return Unknown. Then I updated the patch to modfify OperandInfo to describe a vector register being used as a scalar in 678541701df1c9ca57972f0571c5e0e73cb80dce.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (michaelmaitland) [inline llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1195]: I have to check whether this is the correct operand, or if it should be op3. Will add tests.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1544]: I gave some more thought to this. First I wondered whether it belonged in `isCandidate` since that is the guard to add to the worklist. I don't think it belongs as `isCandidate` since if an MI had VL and SEW operands but did not produce a vector instruction , we would want to reduce it.

For example:
```
a = ..., vl_op, sew_op
b = ..., vl_op, sew_op
scalar_register = some_vector_instr a, b, vl_op, sew_op
```
In this case, we would still want to try to reduce the VL here, despite `some_vector_instr` takes vector instructions and produces a scalar register.

However, I am not aware of any instruction that takes vectors and produces a scalar -- in fact this was explicitly avoided in the case of reductions. It could happen in the future, but I think we don't need to worry about it right now.

With this knowledge, this check is redundant since isCandidate checks for hasVLOp and hasSEWOp and we know that there are no instructions that have this property with a scalar destination. I will move this check to an assert in isCandidate.

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1544]: > However, I am not aware of any instruction that takes vectors and produces a scalar

vmv.x.s and vfmv.f.s

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1544]: Edit: I take that back, VFIRST is an example of such an instruction.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1525]: use -> uses

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1491]: > On a side note, ff the VL is zero, we cannot reduce it any further. We shouldn't even be calling tryReduceVL.

Zero at runtime, not statically known to be zero.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1432]: Drop this debug message. It will print for every instruction in the function which makes it completely uninteresting.

Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1432]: Probably should drop some in the first few bailouts too. And instead add a message when we reach an interesting instruction.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp]: Addressed in 583bb4919602068f9ba02e21503312b372f79ee8.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1487]: Left a FIXME.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1569]: Won't this change the VL to its own VL? What's the point?

Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp]: I removed it.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1567]: Make `checkUsers` return `CanReduceVl`

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (topperc) [inline llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:1560]: Is this comment still relevant?

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:319]: Aren't mask registers always EMUL=1?

> A vector mask occupies only one vector register regardless of SEW and LMUL.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:743]: I saw above that we bail if we have a non-undef passthru, but if we do have a undef passthru then won't operand 1 still be the passthru? And so vs2 will actually be op2

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (topperc) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:319]: Masks still have an effective EMUL within that register. Just like fractional LMUL is only part of a register.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:477]: isOpN was checking if there was a pass through and adjusting by 1 if so. It is more clear that this is happening now that this logic was inlined here.

**Comment by michaelmaitland** [comment]
ping

**Review by preames — APPROVED** [review]
LGTM

As discussed in today's sync call, we have consensus to land this in tree (off by default!) and continue iterating.  There is a good amount of code sharing, and improved testing which needs to happen before this can be enabled by default.  All parties are aware, and enabling that incrementally is the goal of landing now.

**Comment by ppenzin** [comment]
I've ran this by our arch folks and we are also in favor of the change.

**Review by lukel97 — APPROVED** [review]
LGTM to land as is, just spotted something I want to point out before I forget

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:402]: > The signed fractional multiply instruction produces a 2*SEW product of the two SEW inputs, then shifts the result right by SEW-1 bits, rounding these bits according to vxrm, then saturates the result to fit into SEW bits. 

Thats a great catch. I'll add a test specifically for these before merging.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (lukel97) [inline llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:236]: Do we handle compare instructions e.g. vmseq.vv where the destination operand will be a mask register, but not constrained to V0?

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (lukel97) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:402]: Feel free to land first and add a test later, don't let my reviews block this!

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp:236]: from `isMaskOperand`
```
  return Desc.operands()[MO.getOperandNo()].RegClass == RISCV::VMV0RegClassID;
```
It looks like we only support the v0 case for now.

File: llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp
Comment (michaelmaitland) [reply llvm/lib/Target/RISCV/RISCVVLOptimizer.cpp]: i've removed it from getOperandInfo, we can add back with tests later.

**Pull request merged**
